### PR TITLE
Compatible dropwizard metrics

### PR DESCRIPTION
**Pull request by gortiz**
We currently use Yammer as the metric library, but Yammer is not maintained and it doesn’t have new features other more modern metric libraries (like Drowizard or Micrometer) have. For example, Yammer histograms are misleading in some situations (I strongly recommend to read [this article](https://medium.com/expedia-group-tech/your-latency-metrics-could-be-misleading-you-how-hdrhistogram-can-help-9d545b598374) to learn more about the problem). This issue also affects Dropwizard (in fact the referenced article is focused on Dropwizard) but Dropwizard has evolved in a way that the problem can be fixed by changing the reservoir.

We already have a Dropwizard implementation, but we cannot currently use it because it produces metrics with other JMX names, which could break alerts and dashboards our users have right now.

At this moment this PR is a draft on which we can discuss, not a ready to merge PR. This PR changes some default values and therefore we would need to test it properly in environments we can break.

This PR creates a new metric plugin called _compound_. This metric plugin contains a list of other metric plugins. Each time a metric is registered or unregistered in the _compound_ registry, it is registered or unregistered in all other metric plugins. The metric plugins that are notified by the _compound_ registry can be configured, but by default it notifies all other metric plugins in the classpath.

By using this _compound_ metric plugin and including both Dropwizard and Yammer plugins in the classpath, Pinot will register each metric twice: One in Yammer and one in Dropwizard. As said above, each registry produces its own JMX names, so alerts and dashboards created based on the Yammer metrics will continue to work as expected, but given that Dropwizard JMX names will also be published, dashboard and alerts can be migrated to use the new ones and therefore can use the new features.

This PR also adds the ability to change the _domain_ on which Dropwizard metrics are published by changing the property `pinot.metrics.dropwizard.domain`, whose default value is `metrics`, following the Dropwizard default domain. This domain is the prefix used by Dropwizard to create its MBean names, which are something like `"<domain>";type="<type>";name="<metric_name>"`.

By default a Pinot metric called `myTimer` of type `timer` in the Pinot server is exposed:
- by Yammer as `"org.apache.pinot.common.metrics";type="ServerMetrics";name="myTymer"`
- by Dropwizard as `"metrics";type="timers";name="myTymer"`

When the new `pinot.metrics.dropwizard.domain` property is changed to `org.apache.pinot.common.metrics`, Dropwizard will export the metric as `org.apache.pinot.common.metrics;type=timer;name=myTymer`. Note that this name does not double quote the names. For example the domain will be `pinot.metrics.dropwizard.domain`, not `"pinot.metrics.dropwizard.domain"` as it is used in Yammer. As far as I know it is not possible to tell Dropwizard to use other value as `type`. It always use the type of the metric (gauge, timer, histogram, counter, etc).

Therefore the idea would be to do the following:
- Merge this code
- Deploy the resulting image in an environment changing the following properties:
   -  For all X in {server, broker, minion, controller}:
     - `pinot.X.metrics.factory.className` should value `org.apache.pinot.plugin.metrics.compound.CompoundPinotMetricsFactory`.
     - `pinot.X.metrics.dropwizard.domain` to be `org.apache.pinot.common.metrics`
- Analyze the JMX metrics. For each metric we normally have, we should have two:
  - One generated by Yammer called:  `"org.apache.pinot.common.metrics";type="XMetrics";name="whateverMetricName"`
  - Another generated by Dropwizard called: `org.apache.pinot.common.metrics;type=T;name=whateverMetricName`

If we do that, Prometheus should be exporting the Yammer values, as it used to. But we can either use another Prometheus exporter process to check the differences or change the config of the Prometheus agent we usually do to either export both metrics or start exporting using the values from Dropwizard metrics instead of the ones generated by Yammer. The later can be done without changing the Prometheus names and therefore without requiring changes on the alerts or Grafana dashboards.

### CODE REVIEW

**Comment by gortiz** [comment]
Update on the issue:

Metrics created using Dropwizard and Yammer are very close, but not exactly the same. By running using the new CompoundPinotMetricsFactory as factory, we can produce both metrics.
After that, I used [jxmterm](https://cwiki.apache.org/confluence/display/KAFKA/jmxterm+quickstart) to read the JMX metrics exported by Yammer and Dropwizard.

The first difference is the id of the beans created. JMX metrics have several coordinates. Some of them are standard (like domain, name) and some other are not. Specifically, our usage of Yammer and Dropwizard uses these three coordinates:
- domain:
  - Yammer: we set the value to `"org.apache.pinot.common.metrics"` (note> we include the double quotes).
  - Dropwizard: the domain can be changed with `pinot.metrics.dropwizard.domain`. By default it is `metrics`. There is a Javadoc saying that we use that because it was the default in Pinot < 11. I've changed the default to `org.apache.pinot.common.metrics` (without double quotes) to make it closer to Yammer and more difficult to collide with other metrics running in the same system.
- name:
  - Yammer: we wrap the names between double quotes. For example, metric `pinot.server.meetupRsvp_REALTIME.numSegmentsPrunedInvalid` is reported as `"pinot.server.meetupRsvp_REALTIME.numSegmentsPrunedInvalid"` by Yammer
  - Dropwizard: we reported the names without double quotes. For example, metric `pinot.server.meetupRsvp_REALTIME.numSegmentsPrunedInvalid` is reported as `pinot.server.meetupRsvp_REALTIME.numSegmentsPrunedInvalid` by Dropwizard.
- type:
   - Yammer: We set as type either `ControllerMetrics`, `BrokerMetrics`, `MinionMetrics`, `ServerMetrics` or `ValidationMetrics`.
   - Dropwizard: The type is set by Dropwizard, using values like `gauge`, `meters`. It may be possible to change that, but doesn't seem trivial.

As the last part, we configure the prometheus exporter to map JMX metrics to prometheus. In files like `docker/images/pinot/etc/jmx_prometheus_javaagent/configs/broker.yml` (and server.yml, controller.yml, etc) we configure rules that do that mapping. For example:

```yaml
- pattern: "\"org.apache.pinot.common.metrics\"<type=\"BrokerMetrics\", name=\"pinot.broker.([^\\.]*?)\\.totalServerResponseSize\"><>(\\w+)"
  name: "pinot_broker_totalServerResponseSize_$2"
  cache: true
  labels:
    table: "$1"
```

which follows the path: `${domain}<type=${type}, name=${name}`

In order to be able to read Prometheus metrics, we should change the pattern to something like:
```yaml
- pattern: "\"?org.apache.pinot.common.metrics\"?<type=\\w, name=\"pinot.broker.([^\\.]*?)\\.totalServerResponseSize\"><>(\\w+)"
```

Probably we could even skip the type and generate something like:
```yaml
- pattern: "\"?org.apache.pinot.common.metrics\"?<name=\"pinot.broker.([^\\.]*?)\\.totalServerResponseSize\"><>(\\w+)"
```

This changes should not be done when using `CompoundPinotMetricsFactory`, given that it would catch metrics twice (once the one generated by Yammer and once the one generated by Dropwizard)

**Comment by Jackie-Jiang** [comment]
Very promising work! @jackjlli Can you help review this?

**Comment by npawar** [comment]
If we use Compound metrics registry, would the prometheus storage double?
And in the end where you give example of what we can set in pattern with Dropwizard, what should we set to catch separately, if we are using Compound?

**Comment by gortiz** [comment]
> If we use Compound metrics registry, would the prometheus storage double?

That depends on our Prometheus config. We could configure Prometheus to store one set of the metrics or both. In the latter case the storage would be doubled.

> And in the end where you give example of what we can set in pattern with Dropwizard, what should we set to catch separately, if we are using Compound?

I think we should not use Compound in actual deployments. It may be useful to use that in a integration test in order to verify that both metric registries are actually returning compatible information and that we can create the same Prometheus metrics from one or the other.

In case we want to actually use Compound and we want to register both metrics in Prometheus, we would need to have two rules per metric. One can be the same we have right now (which is going to read from Yammer) and then we can add another like:
```
- pattern: "org.apache.pinot.common.metrics<name=\"pinot.broker.([^\\.]*?)\\.totalServerResponseSize\"><>(\\w+)"
```

The latter will only match with Dropwizard because the domain does not start with `\"`

File: pinot-plugins/pinot-metrics/pinot-compound-metrics/src/main/java/org/apache/pinot/plugin/metrics/compound/CompoundPinotMetricsFactory.java
Comment (jackjlli) [inline pinot-plugins/pinot-metrics/pinot-compound-metrics/src/main/java/org/apache/pinot/plugin/metrics/compound/CompoundPinotMetricsFactory.java:43]: Could you add some javadoc (e.g. how to use it and what to note) for this class in case ppl would like to try it out?

Comment (gortiz) [reply pinot-plugins/pinot-metrics/pinot-compound-metrics/src/main/java/org/apache/pinot/plugin/metrics/compound/CompoundPinotMetricsFactory.java:69]: I've added some lines. Is it clearer now?

**Review by jackjlli — COMMENTED** [review]
Overall LGTM. I saw some tests failed. Could you fix them before merging it? Thanks!

**Comment by real-mj-song** [comment]
@gortiz Is this PR still being worked on/active? The compound metrics plugin concept will also help for future open telemetry plugin (when/if it's added).

**Comment by gortiz** [comment]
> Is this PR still being worked on/active? The compound metrics plugin concept will also help for future open telemetry plugin (when/if it's added).

No, it is not. But if it is useful for someone we can add it. Anyway, it would need a +1 from some committer.

**Review by Jackie-Jiang — APPROVED** [review]
Please resolve the conflict

**Comment by gortiz** [comment]
Resolved.

File: pinot-tools/pom.xml
Comment (Jackie-Jiang) [inline pinot-tools/pom.xml:130]: This is defined twice

File: pom.xml
Comment (Jackie-Jiang) [inline pom.xml:418]: This is already defined.

File: pom.xml
Comment (Jackie-Jiang) [inline pom.xml:423]: Can you move this after the `pinot-dropwizard` which is already defined in this file?

**Review by Jackie-Jiang — COMMENTED** [review]
@jackjlli Can you please take another look at this PR?

**Review by jackjlli — APPROVED** [review]
Overall LGTM. Thanks for making the changes in this PR! 👍

**Pull request merged**
